### PR TITLE
WIP: Access and Modify Settings Based on Model Generation

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch changes settings CRUD in state so that model generation is considered when determining which configuration documents are accessed.

## QA steps

TBC

## Documentation changes

Ultimately, yes.

## Bug reference

N/A

### CODE REVIEW

File: state/application.go [hunk 1]
```diff
@@ -709,7 +710,17 @@ func (a *Application) changeCharmOps(
 ) ([]txn.Op, error) {
 	// Build the new application config from what can be used of the old one.
 	var newSettings charm.Settings
-	oldKey, err := readSettings(a.st.db(), settingsC, a.charmConfigKey())
+
```
Comment (manadart) [inline state/application.go]: Changes to the logic accessing the "charms" collection also need to be instituted.

File: state/modelconfig.go [hunk 2]
```diff
@@ -21,11 +22,18 @@ var disallowedModelConfigAttrs = [...]string{
 
 // ModelConfig returns the complete config for the model
 func (m *Model) ModelConfig() (*config.Config, error) {
```
Comment (manadart) [inline state/modelconfig.go]: One wonders why we have identical methods `Config` and `ModelConfig` on `state.Model`.

File: state/modelconfig.go [hunk 3]
```diff
@@ -137,23 +145,30 @@ func (model *Model) modelConfigValues(modelCfg attrValues) (config.ConfigValues,
...
+	}
+	settings, err := readSettings(m.st.db(), globalSettingsC, key, gen == model.GenerationNext)
 	if err != nil {
 		if !errors.IsNotFound(err) {
-			return errors.Annotatef(err, "model %q", model.UUID())
+			return errors.Annotatef(err, "model %q", m.UUID())
 		}
 		// We haven't created settings for this region yet.
-		_, err := createSettings(model.st.db(), globalSettingsC, key, attrs)
+		_, err := createSettings(m.st.db(), globalSettingsC, key, attrs)
```
Comment (manadart) [inline state/modelconfig.go]: This looks wrong.
- The key is unchanged and so will attempt to create a current gen settings doc upon not finding a next gen doc.
- We should probably copy the current doc is it exists in this case anyway to ensure additions/removals are properly handled.

File: state/modelconfig.go [hunk 4]
```diff
@@ -398,7 +417,7 @@ func (st *State) defaultInheritedConfig() (attrValues, error) {
 // controllerInheritedConfig returns the inherited config values
 // sourced from the local cloud config.
 func (st *State) controllerInheritedConfig() (attrValues, error) {
-	settings, err := readSettings(st.db(), globalSettingsC, controllerInheritedSettingsGlobalKey)
+	settings, err := readSettings(st.db(), globalSettingsC, controllerInheritedSettingsGlobalKey, false)
```
Comment (manadart) [inline state/modelconfig.go]: I believe this is correct - such settings are not generational.

File: state/modelconfig.go [hunk 5]
```diff
@@ -422,10 +441,8 @@ func (st *State) regionInheritedConfig(regionSpec *environs.RegionSpec) func() (
 		}
 	}
 	return func() (attrValues, error) {
-		settings, err := readSettings(st.db(),
-			globalSettingsC,
-			regionSettingsGlobalKey(regionSpec.Cloud, regionSpec.Region),
-		)
+		settings, err := readSettings(
```
Comment (manadart) [inline state/modelconfig.go]: Same here.

File: state/modelgeneration.go [hunk 6]
```diff
@@ -306,15 +307,15 @@ func insertGenerationTxnOps(id string) []txn.Op {
...
 	return v, errors.Trace(err)
 }
 
-// ActiveGeneration returns the active generation for the current model.
+// activeGeneration returns the active generation for the current model.
 // If there is no "next" generation pending completion, or if the generation is
 // not active, then the generation to use for config is "current".
 // Otherwise, the model's "next" generation is active.
-func (st *State) ActiveGeneration() (model.GenerationVersion, error) {
+func (st *State) activeGeneration() (model.GenerationVersion, error) {
```
Comment (manadart) [inline state/modelgeneration.go]: The other on-state methods should also be made private like this.

**Comment by manadart** [comment]
[spec card - initial spike target config changes](https://trello.com/c/uvvmeLk9/792-spec-card-initial-spike-target-config-changes)

**Comment by manadart** [comment]
One important outstanding item is to call `AssignApplication` on the generation at the site of config changes made while the "next" generation is active.

File: state/modelconfig.go (hunk 2 above)
Comment (hmlanigan) [reply state/modelconfig.go]: Best guess: the work to "Move model config methods to model receiver."  when state.ModelConfig() was changed to model.ModelConfig(), the existing model.Config() wasn't looked?   The refactoring was related to CaaS implementation. (pr 7827)

Sounds like a tech-debt bug to me.. and maybe a TODO with the bug number, so folks know to use ModelConfig() moving forward?

File: state/model.go [hunk 7]
```diff
@@ -685,7 +687,11 @@ func (m *Model) StatusHistory(filter status.StatusHistoryFilter) ([]status.Statu
 
 // Config returns the config for the model.
 func (m *Model) Config() (*config.Config, error) {
```
Comment (hmlanigan) [inline state/model.go]: How does a generation on a model config work? I wouldn't have expected it based on the spec.

Comment (manadart) [reply state/model.go]: I will seek clarification there. I took it as implied from:

> Which model operations are tracked via generations?
> Not all changes to the model are tracked in this way. 
> - upgrade-charm
> - config
> - attach

File: state/application.go [hunk 8]
```diff
@@ -2078,13 +2119,27 @@ func (a *Application) UpdateCharmConfig(changes charm.Settings) error {
 
 // ApplicationConfig returns the configuration for the application itself.
 func (a *Application) ApplicationConfig() (application.ConfigAttributes, error) {
```
Comment (manadart) [inline state/application.go]: One thing I have missed here is that this has a dependency on if/which unit is requesting config. Just because the next generation is active, doesn't mean the next generation config should be returned for a unit on the current generation.

File: apiserver/facades/client/highavailability/highavailability_test.go [hunk 9]
```diff
@@ -414,31 +415,33 @@ func (s *clientSuite) TestEnableHAPlacementTo(c *gc.C) {
 }
 
 func (s *clientSuite) TestEnableHAPlacementToWithAddressInSpace(c *gc.C) {
-	controllerSettings, _ := s.State.ReadSettings("controllers", "controllerSettings")
+	controllerSettings, _ := s.State.ReadSettings("controllers", "controllerSettings", false)
```
Comment (manadart) [inline apiserver/facades/client/highavailability/highavailability_test.go]: This is the material change. Did the error-checking as a drive-by.

**Comment by mitechie** [comment]
Hello, I am the PR robot. Your PR is older
    than 14 days. I'm closing it for now. If
    it needs attention please work with folks
    to progress the PR and land it.
    Beep..Boop...
